auslib/db.py

```diff
@@ -1189,6 +1252,93 @@ def mergeUpdate(self, old_row, what, changed_by, transaction=None):
+
+    decisionColumns = []
+
+    def __init__(self, db, dialect):
+        self.table.append_column(Column("role", String(50), primary_key=True))
+        self.table.append_column(Column("signoffs_required", Integer, nullable=False))
+
+        super(RequiredSignoffsTable, self).__init__(db, dialect, scheduled_changes=True, scheduled_changes_kwargs={"conditions": ["time"]})
+
+    def getPotentialSignoffs(self, old_row, new_row, transaction=None):
+        potential_signoffs = []
+        for row in (old_row, new_row):
+            if not row:
+                continue
+            where = {col: row[col] for col in self.decisionColumns}
+            potential_signoffs.extend(self.select(where=where, transaction=transaction))
+        return potential_signoffs
+
+    def validate(self, columns, transaction=None):
+        for col in self.decisionColumns:
+            if columns[col] is None:
+                raise ValueError("{} are required.".format(self.decisionColumns))
+
+        users_with_role, = self.db.permissions.user_roles.t.count().where(self.db.permissions.user_roles.role == columns["role"]).execute().fetchone()
+        if users_with_role < columns["signoffs_required"]:
+            msg = ""
+            for col in self.decisionColumns:
+                msg += "{}, ".format(col)
+            raise ValueError("Cannot require {} signoffs for {} - only {} users hold that role".format(columns["signoffs_required"], msg, columns["role"]))
+
+    def insert(self, changed_by, transaction=None, dryrun=False, signoffs=None, **columns):
+        self.validate(columns, transaction=transaction)
+
+        if not self.db.hasPermission(changed_by, "required_signoff", "create", transaction=transaction):
+            raise PermissionDeniedError("{} is not allowed to create new Required Signoffs.".format(changed_by))
+
+        return super(RequiredSignoffsTable, self).insert(changed_by=changed_by, transaction=transaction, dryrun=dryrun, signoffs=signoffs, **columns)
+
+    def update(self, where, what, changed_by, old_data_version, transaction=None, dryrun=False, signoffs=None):
+        for rs in self.select(where=where, transaction=transaction):
+            rs.update(what)
+            self.validate(rs, transaction=transaction)
+
+        if not self.db.hasPermission(changed_by, "required_signoff", "modify", transaction=transaction):
+            raise PermissionDeniedError("{} is not allowed to modify Required Signoffs.".format(changed_by))
+
+        return super(RequiredSignoffsTable, self).update(where=where, what=what, changed_by=changed_by, old_data_version=old_data_version,
+                                                         transaction=transaction, dryrun=dryrun, signoffs=signoffs)
+
+    def delete(self, where, changed_by=None, old_data_version=None, transaction=None, dryrun=False, signoffs=None):
+        if not self.db.hasPermission(changed_by, "required_signoff", "delete", transaction=transaction):
```

Comment:
I don't think that would do anything useful. We always validate before making a change to the database, so we'd just be revalidating the latest database state before deleting it? Unless I'm missing something...